New "Caching" doc page (#1615)
AcademySoftwareFoundation/rez: docs/source/caching.rst

=================
Caching
=================

Resolve Caching
===============

Resolve caching is a feature that caches resolves to a memcached server.

> TODO.

Author: Sweet I added this info thank you!

=======
Caching
=======

Resolve Caching
===============

Resolve caching is a feature that caches resolves to a `memcached <https://memcached.org/>`_ (in-memory) data store.

> Because the data store is in-memory, the full contents of the cache are lost if the memcached service shuts down

Reviewer: I think I would move this in the pros/cons paragraph.

Author: @JeanChristopheMorinPerso Sorry what do you mean?

Reviewer: Sorry, I wasn't very clear. I mean maybe move this to paragraph at line 18?

by any means.

> Memcached is widely used, easy to deploy (because there is no storage needed since it's a single
> process/executable), and therefore only requires a simple server.

Reviewer: ```suggestion
Memcached is a widely used, easy to deploy single process stateless server. It uses very little resources and it's very fast.
```

In a studio environment (with many machines), machines that perform a solve that is already cached to the
resolve cache will simply receive the cached result rather than preforming a re-solve.

Resolve caching has almost no downsides. Only in rare edge cases where you have to "hack" a released package into
production do you see any issues. In this case, because resolves are cached, you may receive a different package than
you expect. In this case however, it's better to just manually invalidate the cache anyway.

Cache contents
--------------
The following information is stored to the memcached server for each solve:

* Solver information about the previously cached solve.
* Release times information about when each package variant in the resolve was last released.
* Variant states information about the state of a variant. For example, in the 'filesystem' repository type, the 'state' is the last modified date of the file associated with the variant (perhaps a package.py). If the state of any variant has changed from a cached resolve - eg. if a file has been modified - the cached resolve is discarded.

Setup
-----

> To enable memcached caching, you need to configure the :data:`memcached_uri` config variable. This variable accepts a list of memcached uri servers or None. Example with memcached running on localhost on its default port:

Reviewer: ```suggestion
To enable memcached caching, you need to configure the :data:`memcached_uri` config variable. This variable accepts a list of URI to your memcached servers or None. Example with memcached running on localhost on its default port:
```

> =================
> Caching
> =================

Reviewer: ```suggestion
=======
Caching
=======
```

Resolve Caching
===============

> Resolve caching is a feature that caches resolves to a memcached (in-memory) server. Because the server is in-memory,

Reviewer: ```suggestion
Resolve caching is a feature that caches resolves to a `memcached <https://memcached.org/>`_ (in-memory) server. Because the server is in-memory,
```
I think we can probably merge the "benefits" and downsides" into this section. Also the server isn't in-memory. It's the data it holds that is in memory. I would also mention that memcached is a widely used, easy to deploy (because there is no storage needed and because it's a single process/executable) and simple server.

the full contents of the cache are lost if the memcached service shuts down by any means.

Cache contents
--------------
The following information is stored to the memcached server for each solve:

* Solver information about the previously cached solve.
* Release times information about when each package variant in the resolve was last released.
* Variant states information about the state of a variant. For example, in the 'filesystem' repository type, the 'state' is the last modified date of the file associated with the variant (perhaps a package.py). If the state of any variant has changed from a cached resolve - eg. if a file has been modified - the cached resolve is discarded.

Setup
-----

To enable memcached caching, you need to configure the :data:`memcached_uri` config variable. This variable accepts a list of memcached uri servers or None. Example with memcached running on localhost on its default port:

.. code-block:: console

   memcached_uri = ["127.0.0.1:11211"]

This is the only parameter you need to configure to enable caching of the content and location of package file definitions and resolutions in Rez.

> Configuration
> -------------
> There are several variables that can be accessed to modify the default behavior of resolve caching:
> 
> * :data:`resolve_caching`
> * :data:`cache_package_files`
> * :data:`cache_listdir`
> * :data:`resource_caching_maxsize`
> * :data:`memcached_package_file_min_compress_len`
> * :data:`memcached_context_file_min_compress_len`
> * :data:`memcached_listdir_min_compress_len`
> * :data:`memcached_resolve_min_compress_len`

Reviewer: ```suggestion
Please refer to the :ref:`caching` configuration section for a complete list of settings.
```
?

Author: @JeanChristopheMorinPerso To make sure I'm understanding this correctly, the `:ref` is just a cross reference directive right? https://www.sphinx-doc.org/en/master/usage/referencing.html#role-ref

If so, this doesn't seem to be working for some reason. Maybe because the new page's heading is also "Caching"? I'm not sure.

> Validating operation

Reviewer: I think I would move this section after `Cache invalidation`.

--------------------------
To print debugging information about memcached usage, you can temporarily declare the following variables in a terminal:

.. code-block:: console

   export REZ_DEBUG_MEMCACHE=1 (linux/macos bash)
   $env:REZ_DEBUG_MEMCACHE=1 (powershell)

or set :data:`debug_memcache` to True in your rezconfig.py.

Cache invalidation

> ----------------------

Reviewer: ```suggestion
------------------
```

Cache entries will automatically be invalidated when a newer package version is released that would change the result
of an existing resolve.

For example, let's say you are running rez-env with the package ``foo1+<2``, and originally, the only available
``foo`` package version is ``1.0.0``, so the cached resolve points to ``1.0.0``. However, at some point afterwards
you release a new version ``1.0.1``. The cache would invalidate for the request ``foo1+<2`` and the next resolve
would correctly retrieve package version ``1.0.1``.

Show stats from memcached server
--------------------------------

> Rez provides a command-line tool :ref:`rez-memcache` for query the memcached server and obtaining status information and statistics.

Reviewer: ```suggestion
Rez provides a command-line tool :ref:`rez-memcache` that can be used to see stats about cache misses/hits and to reset the memcached cache.
```

.. code-block:: console

   $ rez-memcache

   CACHE SERVER               UPTIME      HITS      MISSES  HIT RATIO  MEMORY  USED
   ------------               ------      ----      ------  ---------  ------  ----
   127.0.0.1:11211            20 hours    27690     5205    84%        119 Gb  10 Mb (0%)
   central.example.com:11211  6.2 months  19145089  456     99%        64 Mb   1.9 Mb (2%)

> Benefits

Reviewer: I would put both the benefits and downside right at the top, right under "Resolve Caching". I would also not make them sections.

--------
In a studio environment (with many machines), machines that perform a solve that is already cached to the
resolve cache will simply receive the cached result rather than preforming a re-solve.

Downsides
---------
Resolve caching has almost no downsides. Only in rare edge cases where you have to "hack" a released package into
production do you see any issues. In this case, because resolves are cached, you may receive a different package than
you expect. In this case however, it's better to just manually invalidate the cache anyway.

.. _package-caching:

Package Caching
===============

Package caching is a feature that copies package payloads onto local disk in
order to speed up runtime environments. For example, if your released packages
reside on shared storage (which is common), then running say, a Python process,

> will fetch all source from the shared storage across your network. The point of

Reviewer: ```suggestion
will load all source from the shared storage across your network. The point of
```

the cache is to copy that content locally instead, and avoid the network cost.

.. note::

>    Please note: Package caching does **NOT** cache package

Reviewer: ```suggestion
   Package caching does **NOT** cache package
```
